File: ParteJava/ProyectoPro3/ProyectoPro3Persistencia/src/main/java/pe/edu/pucp/proyectopro3/dao/reportes/ReporteReclamoDAOImpl.java
package pe.edu.pucp.proyectopro3.dao.reportes;

import pe.edu.pucp.proyectopro3.dao.DefaultBaseDAO;
import pe.edu.pucp.proyectopro3.modelo.reportes.ReporteReclamo;

import java.sql.*;

public class ReporteReclamoDAOImpl extends DefaultBaseDAO<ReporteReclamo> implements ReporteReclamoDAO {
    private static final int DEFAULT_ID_USUARIO = 1;

    @Override
    protected PreparedStatement comandoCrear(Connection conn, ReporteReclamo modelo) throws SQLException {
        String sql = "{call sp_InsertReporte_Reclamo(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        cmd.setDate(1, new java.sql.Date(modelo.getFechaGeneracion().getTime()));
        cmd.setDate(2, new java.sql.Date(modelo.getFechaInicioFiltro().getTime()));
        cmd.setDate(3, new java.sql.Date(modelo.getFechaFinFiltro().getTime()));
        cmd.setInt(4, modelo.getCantidadReservas());
        cmd.setInt(5, modelo.getCantidadReclamos());
        cmd.setDouble(6, modelo.getPorcentajeIncidencias());
        cmd.setInt(7, modelo.getTotalProcede());
        cmd.setInt(8, modelo.getTotalNoProcede());
        cmd.setInt(9, modelo.getTotalPendientes());
        cmd.setInt(10, DEFAULT_ID_USUARIO);
        cmd.registerOutParameter(11, Types.INTEGER);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, ReporteReclamo modelo) throws SQLException {
        String sql = "UPDATE Reporte_Reclamo SET fecha_generacion=?, fecha_inicio_filtro=?, "
                + "fecha_fin_filtro=?, cantidad_reservas=?, cantidad_reclamos=?, "
                + "porcentaje_incidencias=?, total_procede=?, total_no_procede=?, "
                + "total_pendientes=? WHERE id_reporte_reclamo=?";
        PreparedStatement cmd = conn.prepareStatement(sql);

        cmd.setDate(1, new java.sql.Date(modelo.getFechaGeneracion().getTime()));
        cmd.setDate(2, new java.sql.Date(modelo.getFechaInicioFiltro().getTime()));
        cmd.setDate(3, new java.sql.Date(modelo.getFechaFinFiltro().getTime()));
        cmd.setInt(4, modelo.getCantidadReservas());
        cmd.setInt(5, modelo.getCantidadReclamos());
        cmd.setDouble(6, modelo.getPorcentajeIncidencias());
        cmd.setInt(7, modelo.getTotalProcede());
        cmd.setInt(8, modelo.getTotalNoProcede());
        cmd.setInt(9, modelo.getTotalPendientes());
        cmd.setInt(10, modelo.getIdReporte());

        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_DeleteReporte_Reclamo(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_ListReporteReclamoById(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call spListReporteReclamo()}";
        return conn.prepareCall(sql);
    }

    @Override
    protected Integer extraerIdDesdeCallable(CallableStatement cmd) throws SQLException {
        return cmd.getInt(11);
    }

    @Override
    protected Integer extraerIdDesdeGeneratedKeys(ResultSet rs) throws SQLException {
        return rs.getInt(1);
    }

    @Override
    protected ReporteReclamo mapearModelo(ResultSet rs) throws SQLException {
        ReporteReclamo reporte = new ReporteReclamo();

        reporte.setIdReporte(rs.getInt("id_reporte_reclamo"));
        reporte.setFechaGeneracion(rs.getDate("fecha_generacion"));
        reporte.setFechaInicioFiltro(rs.getDate("fecha_inicio_filtro"));
        reporte.setFechaFinFiltro(rs.getDate("fecha_fin_filtro"));
        reporte.setCantidadReservas(rs.getInt("cantidad_reservas"));
        reporte.setCantidadReclamos(rs.getInt("cantidad_reclamos"));
        reporte.setPorcentajeIncidencias(rs.getDouble("porcentaje_incidencias"));
        reporte.setTotalProcede(rs.getInt("total_procede"));
        reporte.setTotalNoProcede(rs.getInt("total_no_procede"));
        reporte.setTotalPendientes(rs.getInt("total_pendientes"));

        return reporte;
    }
}

File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Persistencia/DAO/reportes/ReporteReclamoDaoImpl.cs
using System.Data;
using System.Data.Common;
using ProyectoProgramacion3Model.Model.reportes;

namespace ProyectoProgramacion3Persistencia.DAO.reportes;

public class ReporteReclamoDaoImpl : DefaultBaseDao<ReporteReclamo>, IReporteReclamoDao
{
    private const int DefaultIdUsuario = 1;

    protected override DbCommand ComandoCrear(DbConnection conn, ReporteReclamo modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_InsertReporte_Reclamo");
        CrearParametro(cmd, "p_fecha_generacion", DateOnlyParam(modelo.fechaGeneracion), DbType.Date);
        CrearParametro(cmd, "p_fecha_inicio_filtro", DateOnlyParam(modelo.fechaInicioFiltro), DbType.Date);
        CrearParametro(cmd, "p_fecha_fin_filtro", DateOnlyParam(modelo.fechaFinFiltro), DbType.Date);
        CrearParametro(cmd, "p_cantidad_reservas", modelo.cantidadReservas);
        CrearParametro(cmd, "p_cantidad_reclamos", modelo.cantidadReclamos);
        CrearParametro(cmd, "p_porcentaje_incidencias", modelo.porcentajeIncidencias);
        CrearParametro(cmd, "p_total_procede", modelo.totalProcede);
        CrearParametro(cmd, "p_total_no_procede", modelo.totalNoProcede);
        CrearParametro(cmd, "p_total_pendientes", modelo.totalPendientes);
        CrearParametro(cmd, "p_id_usuario", DefaultIdUsuario);
        CrearParametroSalida(cmd, "_id_generado", DbType.Int32);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, ReporteReclamo modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "UPDATE Reporte_Reclamo SET fecha_generacion=@fechaGeneracion, fecha_inicio_filtro=@fechaInicio, fecha_fin_filtro=@fechaFin, cantidad_reservas=@reservas, cantidad_reclamos=@reclamos, porcentaje_incidencias=@porcentaje, total_procede=@procede, total_no_procede=@noProcede, total_pendientes=@pendientes WHERE id_reporte_reclamo=@id";
        CrearParametro(cmd, "@fechaGeneracion", DateOnlyParam(modelo.fechaGeneracion), DbType.Date);
        CrearParametro(cmd, "@fechaInicio", DateOnlyParam(modelo.fechaInicioFiltro), DbType.Date);
        CrearParametro(cmd, "@fechaFin", DateOnlyParam(modelo.fechaFinFiltro), DbType.Date);
        CrearParametro(cmd, "@reservas", modelo.cantidadReservas);
        CrearParametro(cmd, "@reclamos", modelo.cantidadReclamos);
        CrearParametro(cmd, "@porcentaje", modelo.porcentajeIncidencias);
        CrearParametro(cmd, "@procede", modelo.totalProcede);
        CrearParametro(cmd, "@noProcede", modelo.totalNoProcede);
        CrearParametro(cmd, "@pendientes", modelo.totalPendientes);
        CrearParametro(cmd, "@id", modelo.idReporte);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_DeleteReporte_Reclamo");
        CrearParametro(cmd, "p_id_reporte_reclamo", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_ListReporteReclamoById");
        CrearParametro(cmd, "p_id_reporte_reclamo", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        return CrearStoredProcedure(conn, "spListReporteReclamo");
    }

    protected override ReporteReclamo MapearModelo(DbDataReader reader)
    {
        return new ReporteReclamo
        {
            idReporte = LeerEntero(reader, "id_reporte_reclamo"),
            fechaGeneracion = LeerDateOnly(reader, "fecha_generacion"),
            fechaInicioFiltro = LeerDateOnly(reader, "fecha_inicio_filtro"),
            fechaFinFiltro = LeerDateOnly(reader, "fecha_fin_filtro"),
            cantidadReservas = LeerEntero(reader, "cantidad_reservas"),
            cantidadReclamos = LeerEntero(reader, "cantidad_reclamos"),
            porcentajeIncidencias = LeerDouble(reader, "porcentaje_incidencias"),
            totalProcede = LeerEntero(reader, "total_procede"),
            totalNoProcede = LeerEntero(reader, "total_no_procede"),
            totalPendientes = LeerEntero(reader, "total_pendientes")
        };
    }
}
